File: solutions/simulation/typescript/1275-find-winner-on-a-tic-tac-toe-game.ts
/**
 * # 1275. Find Winner On A Tic Tac Toe Game
 *
 * LeetCode Problem 1275: Find Winner on a Tic Tac Toe Game
 * Difficulty: Easy
 * Category: Simulation
 *
 * METADATA:
 * Techniques: Simulation, Game state checking
 * Data Structures: 2D array
 * Patterns: Win condition checking
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

function tictactoe(moves: number[][]): string {
    const board: string[][] = Array(3).fill(null).map(() => Array(3).fill(''));

    for (let i = 0; i < moves.length; i++) {
        const [row, col] = moves[i];
        board[row][col] = i % 2 === 0 ? 'X' : 'O';
    }

    function checkWinner(player: string): boolean {
        // Check rows
        for (const row of board) {
            if (row.every(cell => cell === player)) return true;
        }
        // Check columns
        for (let col = 0; col < 3; col++) {
            if (board.every(row => row[col] === player)) return true;
        }
        // Check diagonals
        if (board[0][0] === player && board[1][1] === player && board[2][2] === player) return true;
        if (board[0][2] === player && board[1][1] === player && board[2][0] === player) return true;
        return false;
    }

    if (checkWinner('X')) return "A";
    if (checkWinner('O')) return "B";
    if (moves.length === 9) return "Draw";
    return "Pending";
}

if (require.main === module) {
    const testCases: [number[][], string][] = [
        [[[0, 0], [2, 0], [1, 1], [2, 1], [2, 2]], "A"],
        [[[0, 0], [1, 1], [0, 1], [0, 2], [1, 0], [2, 0]], "B"],
        [[[0, 0], [1, 1], [2, 0], [1, 0], [1, 2], [2, 1], [0, 1], [0, 2], [2, 2]], "Draw"],
    ];

    console.log("Testing tictactoe:");
    for (const [moves, expected] of testCases) {
        const result: string = tictactoe(moves);
        const status: string = result === expected ? "✓" : "✗";
        console.log(`${status} tictactoe(${JSON.stringify(moves)}) = ${result}, expected = ${expected}`);
    }
}

export { tictactoe };

File: solutions/simulation/python/1275-find-winner-on-a-tic-tac-toe-game.py
"""### METADATA:

### INTUITION:
The key insight is to solve this problem efficiently.
 *

### APPROACH:
1. **Initialize data structures**: Set up the required data structures for the algorithm
2. **Process input**: Iterate through the input applying the core technique
3. **Track state**: Maintain necessary state information during processing
4. **Return result**: Construct and return the final solution

### WHY THIS WORKS:
This approach works because it correctly implements the problem requirements.
 *

### EXAMPLE WALKTHROUGH:
Input:
```
example input
```

Output:
```
example output
```

Step-by-step execution:
1. [First step]
2. [Second step]
3. [Final step]

### TIME COMPLEXITY:
**O(n²)** - Analysis of time complexity
 * - [Add explanation of why this complexity]

### SPACE COMPLEXITY:
**O(1)** - Analysis of space complexity
 * - [Add explanation of why this complexity]

### EDGE CASES:
- **Empty input**: Handle when input is empty
- **Single element**: Handle single-element inputs
- **Boundary values**: Handle minimum/maximum valid values

"""

from typing import List


def tictactoe(moves: List[List[int]]) -> str:
    """Determine tic-tac-toe game winner."""
    board = [['' for _ in range(3)] for _ in range(3)]

    # Place moves alternating between A and B
    for i, (row, col) in enumerate(moves):
        board[row][col] = 'X' if i % 2 == 0 else 'O'

    # Check winner
    def check_winner(player: str) -> bool:
        """Check if player has won."""
        # Check rows
        for row in board:
            if all(cell == player for cell in row):
                return True
        # Check columns
        for col in range(3):
            if all(board[row][col] == player for row in range(3)):
                return True
        # Check diagonals
        if all(board[i][i] == player for i in range(3)):
            return True
        if all(board[i][2 - i] == player for i in range(3)):
            return True
        return False

    if check_winner('X'):
        return "A"
    if check_winner('O'):
        return "B"
    if len(moves) == 9:
        return "Draw"
    return "Pending"


if __name__ == "__main__":
    test_cases = [
        ([[0, 0], [2, 0], [1, 1], [2, 1], [2, 2]], "A"),
        ([[0, 0], [1, 1], [0, 1], [0, 2], [1, 0], [2, 0]], "B"),
        ([[0, 0], [1, 1], [2, 0], [1, 0], [1, 2], [2, 1], [0, 1], [0, 2], [2, 2]], "Draw"),
        ([[0, 0], [1, 1]], "Pending"),
    ]

    print("Testing tictactoe:")
    for moves, expected in test_cases:
        result = tictactoe(moves)
        status = "✓" if result == expected else "✗"
        print(f"{status} tictactoe({moves}) = result, expected = expected")
